### DISCUSSION TITLE
Generated fields and create command

### DISCUSSION THREAD

**Discussion by dzenana-s** [sensedeep/dynamodb-onetable#381] [General]
We are using two generated time fields: `createdAt` and `updatedAt`. With version `2.4.0`, I think because of this change https://github.com/sensedeep/dynamodb-onetable/commit/22c21291f85a53c3bc1e224bc409cddd871cf17e to lines 274, we are getting an error because we do not send these values on create. Is it anticipated that generated fields should also be supplied on create? Should we send any info and it will do the generation anyways? Please advise. 

It looks like the change to the `Model.d.ts` file right before it is the one that did add the option of omitting generated fields. But both seem to have gone on to version `2.4.0`, wondering what is the anticipated method for generated fields to be handled.

**Comment by mobsense** [comment]
You should not need to provide them and you should not need them in your schema as we add them internally.

I've checked this and it works for me.

Can you create a test case that demonstrates?

**Reply by dzenana-s** [reply]
Yes, I should clarify a bit. It provides a Typescript type error - if I send to create something like this: 

```
const Schema = {
    version: '0.0.1',
    indexes: {
        primary: { hash: 'pk', sort: 'sk' },
    },
    models: {
        User: {
            pk:    { type: String, value: '${_type}#${id}' },
            sk:    { type: String, value: '${_type}#' },
            id:    { type: String, generate: 'ulid' },
            name:  { type: String },
            email: { type: String },
            createdAt: { type: Date, required: true },
            updatedAt: { type: Date, required: true },
        }
    },
    params: {
        timestamps: true,
        createdField: 'createdAt',
        updatedField: 'updatedAt',
    },
}
export type UserType = Entity<typeof Schema.models.User>;

const table = new Table({
    name: 'TimestampsTable',
    client: Client,
    schema: Schema
}) 
const User = table.getModel('User')

// The parameters we create with should be without these date fields 
const userProperties: Omit<UserType, 'createdAt' | 'updatedAt'> = {
    pk: "pk",
    sk: "sk", 
    id: "id", 
    name: "Bob", 
    email: "bob@gmail.com"
}
// The following line a typescript type error like: 
// error TS2345: Argument of type 'Omit<UserType, "createdAt" | "updatedAt">' is not assignable to parameter of type 'UserType'. 
resultingUser: UserType = User.create(userProperties) 

```

I think this is because line 247 in this change https://github.com/sensedeep/dynamodb-onetable/commit/22c21291f85a53c3bc1e224bc409cddd871cf17e makes it so that 
`create(properties: T, params?: OneParams): Promise<T>;` so the type of what comes in and is returned should be the same. It previously was `create(properties: CreateProperties<T>, params?: OneParams): Promise<T>;` so it worked to provide only a portion of the properties in the create as will be returned.

**Comment by mobsense** [comment]
Ah, thank you. That helps.

I need to be more specific when asking did the runtime error, or did the TS error. In this case, I thought you were referring to a runtime error and not a TS error.

What happens if you remove the required from the schema?  OneTable will still populate the values internally.

**Reply by dzenana-s** [reply]
Thank you for the prompt responses! 

When I remove it from the schema, I get this type of type error: 

```
error TS2322: Type '{ id: string; (MORE_FIELDS_OMITTED_FOR_READING), createdAt: Date; updatedAt: Date; }' is not assignable to type 'AccountConfigType'.
Object literal may only specify known properties, and 'createdAt' does not exist in type 'AccountConfigType'.
```

**Comment by mobsense** [comment]
For TS, it is best to include createdAt, updatedAt in the schema, BUT not make them required.

Required means you MUST provide them to the create API. As OneTable creates them for you, you don't have to provide them.
